Packages/ClientRuntime/Sources/Waiters/WaiterRetryer.swift

```diff
@@ -0,0 +1,112 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// The interface for the type which delays per the wait schedule, performs the request, and evaluates the result.
+/// Called by the `Waiter` to make the initial request, then called again for an additional request when retry
+/// is needed.
+/// Having a retryer protocol allows `WaiterRetryer` to be mocked for testing.
+protocol WaiterRetryerInterface {
+    associatedtype RetryerInput
+    associatedtype RetryerOutput
+
+    ///
+    func waitThenRequest(
+        scheduler: WaiterScheduler,
+        input: RetryerInput,
+        config: WaiterConfiguration<RetryerInput, RetryerOutput>,
+        operation: (RetryerInput) async throws -> RetryerOutput
+    ) async throws -> WaiterOutcome<RetryerOutput>?
+}
+
+/// Container used to type-erase the retryer which implements `WaiterRetryerInterface`
+/// Used to permit injection of a `WaiterRetryerMock` when testing `Waiter`.
+class WaiterRetryerContainer<Input, Output>: WaiterRetryerInterface {
```

Comment:
Can we remove this class in favor of reworking `WaiterRetryer` to be more like this container aka more composable? This would prevent our testability things from leaking into our implementation. Additionally, it'll make things a bit simpler by no longer needing two constructs with similar names and purposes (WaiterRetryer + WaiterRetryerContainer) and we no longer need the protocol either.

```swift
struct WaiterRetryer<Input, Output> {
    typealias Operation = (Input) async throws -> Output
    typealias WaitThenRequest = (WaiterScheduler, Input, WaiterConfiguration<Input, Output>, Operation) async throws -> WaiterOutcome<Output>?
    
    let waitThenRequest: WaitThenRequest
    
    // Allow for folks to inject in their own implementation for `waitThenRequest` but default to our "live" implementation is nothing is provided
    init(waitThenRequest: @escaping WaitThenRequest = Self.smithy(scheduler:input:config:operation:)) {
        self.waitThenRequest = waitThenRequest
    }
    
    private static func smithy(
        scheduler: WaiterScheduler,
        input: Input,
        config: WaiterConfiguration<Input, Output>,
        operation: Operation
    ) async throws -> WaiterOutcome<Output>? {
        // Find the required delay from the scheduler, and wait if required.
        if scheduler.currentDelay > 0.0 {
            try await Task.sleep(nanoseconds: UInt64(scheduler.currentDelay * 1_000_000_000.0))
        }

        // Try the operation, collect the result & update the scheduler (for test use only) for this attempt.
        let result = await Result<Output, Error>(catching: { try await operation(input) })
        scheduler.updateAfterRetry()

        // Test the acceptors, to see if one matches.  Take the first match if there is one.
        // If a match is found, take action as required.
        if let match = config.acceptors.compactMap({ $0.evaluate(input: input, result: result) }).first {
            switch match {
            case .success(let lastResult):
                return WaiterOutcome(attempts: scheduler.attempt, result: lastResult)
            case .failure(let lastResult):
                throw WaiterFailureError<Output>(attempts: scheduler.attempt, failedOnMatch: true, result: lastResult)
            case .retry:
                return nil  // Returning nil causes retry, if time remains
            }
        // If no matching acceptor is found, fail if the result was an error.
        } else if case .failure(let error) = result {
            throw WaiterFailureError<Output>(
                attempts: scheduler.attempt,
                failedOnMatch: false,
                result: .failure(error)
            )
        }
        // Since no acceptor matched and the response was not an unhandled error,
        // return nil to cause retry, if time remains
        return nil
    }
}
```